flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/diff/SpecDiffTest.java

```diff
@@ -0,0 +1,171 @@
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Stream;
+
+import static org.apache.flink.kubernetes.operator.config.KubernetesOperatorConfigOptions.OPERATOR_RECONCILE_INTERVAL;
+import static org.apache.flink.kubernetes.operator.metrics.KubernetesOperatorMetricOptions.SCOPE_NAMING_KUBERNETES_OPERATOR;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
+
+/** Spec diff test. */
+public class SpecDiffTest {
+
+    @ParameterizedTest
+    @MethodSource("applicationTestParams")
+    public void testIgnore(AbstractFlinkSpec left, AbstractFlinkSpec right) {
+        var diff = left.diff(right);
+        assertEquals(DiffType.IGNORE, diff.getType());
+        assertEquals(0, diff.getNumDiffs());
+        if (right.getJob() != null) {
+            right.getJob().setUpgradeMode(UpgradeMode.LAST_STATE);
+            right.getJob().setAllowNonRestoredState(true);
+            right.getJob().setInitialSavepointPath("local:///tmp");
+            right.getJob().setSavepointTriggerNonce(123L);
+            diff = left.diff(right);
+            assertEquals(DiffType.IGNORE, diff.getType());
+            assertEquals(4, diff.getNumDiffs());
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("applicationTestParams")
+    public void testScale(AbstractFlinkSpec left, AbstractFlinkSpec right) {
+        var diff = left.diff(right);
+        assertEquals(DiffType.IGNORE, diff.getType());
+        assertEquals(0, diff.getNumDiffs());
+
+        if (right.getJob() != null) {
+            right.getJob().setParallelism(100);
+            diff = left.diff(right);
+            assertEquals(DiffType.SCALE, diff.getType());
+            assertEquals(1, diff.getNumDiffs());
+
+            right.getJob().setUpgradeMode(UpgradeMode.LAST_STATE);
+            right.getJob().setAllowNonRestoredState(true);
+            right.getJob().setInitialSavepointPath("local:///tmp");
+            right.getJob().setSavepointTriggerNonce(123L);
+            diff = left.diff(right);
+            assertEquals(DiffType.SCALE, diff.getType());
+            assertEquals(5, diff.getNumDiffs());
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("applicationTestParams")
+    public void testReconcile(AbstractFlinkSpec left, AbstractFlinkSpec right) {
+        var diff = left.diff(right);
+        assertEquals(DiffType.IGNORE, diff.getType());
+        assertEquals(0, diff.getNumDiffs());
+
+        if (left instanceof FlinkDeploymentSpec) {
+            ((FlinkDeploymentSpec) right).setImage("flink:greatest");
+            assertEquals(DiffType.UPGRADE, left.diff(right).getType());
+            diff = left.diff(right);
+            assertEquals(DiffType.UPGRADE, diff.getType());
+            assertEquals(1, diff.getNumDiffs());
+        }
+
```

Comment:
It's a bit strange to use `ParameterizedTest` here because you actually have 3 different inner ifs for the 3 different params. I would personally prefer to have single test cases because the parameterization just breaks up the logic  (hides what is tested) and its harder to understand.

Reply:
Dropped the `ParameterizedTest` approach